lib/solidus_importer/configuration.rb

```diff
@@ -51,6 +51,21 @@ def available_types
     # argument. Using a lambda will verify that the correct number of arguments
     # are passed, whereas using a plain Proc won't.
     preference :row_exception_handler, :callable, default: proc { ->(exception, context) {} }
+
+    # An array of callables that will validate the imported data after it has
+    # been parsed and before it is imported. A validation message should be
+    # returned on failure, otherwise `nil` should be returned on success.
+    # A `CSV::Table` object will be passed to the callable.
+    # e.g.
+    # SolidusImporter::Config.csv_format_validators = [
+    #   ->(csv_table) do
+    #     headers = csv_table.headers
+    #     if headers.blank? || !headers.exclude?(nil)
+    #       'Invalid headers'
+    #     end
+    #   end
+    # ]
+    preference :csv_format_validators, :array, default: []
```

Comment:
I'm concerned that someone could look at this `default` value here and miss that we are actually setting the default in the template. Why did you go in this direction? Is this way of setting default something idiomatic seen in other extensions? 

I saw in the (fantastic) commit description that the default is set to avoid breaking change but I'm afraid this is still breaking because the template file is not copied again when the extension is updated (IIRC).

Reply:
I went in this direction to make it easier to remove the default validation, and add awareness to new default validations that may get added in future releases! Going to 

if you set the default in the `lib/solidus_importer/configuration.rb` file, then in order to remove the existing address validator, you have to totally blow away the existing value.

```rb
config.import_data_validators = [ ]

config.import_data_validators << your_custom_validator
```

Now, let's say in a future release, another default validator is added to `lib/solidus_importer/configuration.rb`
After the user upgrades this gem, they are now blowing away **both** of the default validators, even though they may have only **intended** to remove the header validator in the first place.

If the default is configured in the initializer, the user has more visibility over changes to default validators when they upgrade. However, this depends on the assumption that when users are upgrading solidus extensions, they **re-run the install generators after updating the gem.** This is something I personally have gotten in the habit of doing as I've used extensions that introduce required options/changes into the initializer files, but I don't know if it's idiomatic/standard practice.

If you have overridden the `config.import_data_validators` value in the initializer, and you run the install generators after an upgrade, solidus dev support will try to override the `initializer.rb file with the new defaults. The user can then compare the new defaults with their initializer override and decide what settings they want to keep (i.e. maybe they want to keep the new default validator but still omit the header validator) 

Solidus does something similar but more refined - it adds a `solidus:update` generator which will generate an initializer file that shows the changed defaults.
https://edgeguides.solidus.io/getting-started/upgrading-solidus#updating-preferences

I don't think this is idiomatic, and in reflection, I think if we go down this route, an `upgrading instructions` section should be added to the readme.

Reply:
Curious on your thoughts and happy to change it to being set in the `lib/solidus_importer/configuration.rb` file!

Also had the thought that if the header validator wasn't a lambda but a callable class instead, then the user could explicitly find and remove it in their initializer override.

```rb
config.import_data_validators.filter do |validator| validator.class.name != SolidusImporter::HeaderValidator end

config.import_data_validators << your_custom_validator
```

Reply:
👍 works for me as is as soon as we add that part in the README. What about also adding a code comment in the configuration file to point to where the actual default is set?

> Also had the thought that if the header validator wasn't a lambda but a callable class instead, then the user could explicitly find and remove it in their initializer override.

Interesting, but if we go with the template solution, I think users will have enough flexibility to do what they want. Still, probably a list of classes would be more readable/enjoyable to work with than a list of lambdas. Open to seeing some changes here, even in another PR.

Reply:
Sweet!

I added a section on upgrading to the readme, and added a code comment to the configuration file to point to the initializer!

Reply:
I also just ended up added a concrete callable class for the header validator to make the configuration file easier to read than lambas!